### ISSUE TITLE
A standard concept of "user session id"

### ISSUE THREAD

**Issue by hamstar0** [tModLoader/tModLoader#1525]
### Description
Have an implementation of a standard concept of a [session id](https://en.wikipedia.org/wiki/Session_ID) for a standard way to recognize users on a server uniquely. This ID is not known to anyone else but the current user and the server.

Context: <https://discord.com/channels/103110554649894912/445276626352209920/842889069809631243>

### What does this proposal attempt to solve or improve?
Allows for recognizing users by servers in a secure way (such as to give them privileges) that can be remembered between gameplay sessions. Additional code would be needed to implement proper logins for a formal concept of users, but that would still depend on the presence of session ids to map them to (i.e. upon a login authentication).

### Which (other) solutions should be considered?
If Steam IDs were available somehow for GoG clients, and also if other clients were not made aware of a given user's Steam ID on a server, these might work.

**Comment by Mirsario** [comment]
The Main.clientUUIID string seems completely satisfactory. Just need to store it on players in MessageBuffer, as the packet with the UUID is currently ignored by servers:
![image](https://user-images.githubusercontent.com/6795251/118410749-be117a00-b699-11eb-8434-5a05c7baa668.png)

**Comment by hamstar0** [comment]
Yeah, and it should be mapped to each player.

**Comment by Killia0** [comment]
> The Main.clientUUIID string seems completely satisfactory. Just need to store it somewhere after MessageBuffer receives the packet that contains a player's UUID. It is currently ignored by servers:
> ![image](https://user-images.githubusercontent.com/6795251/118410749-be117a00-b699-11eb-8434-5a05c7baa668.png)

Hello, I don't really deal with tmodloader besides recently looking at the github progress to see if I want to delve into it (probably once stable 1.4 is out). But I do deal with servers a lot (On a big vanilla TShock server) so I felt like giving my opinion on this. 
I wouldn't say `Main.clientUUID` is satisfactory as a reliable way to identify a player; it is something that can really easily be changed (it's just a field on the .json config, with a simple GUID format), which is an issue when either a player wants to be malicious on your server, or just if you rely on this ID for any persistent mapping of values and the player accidentally deleted and reset their config. Ideally you'd have something that absolutely tracks a unique player/game installation that can't be easily spoofed (or well for basic purposes something that wouldn't change on accident) but I'm not sure how you'd want to handle this with steamID not being something every installation has, maybe mac adress as an alternative?.

**Comment by Chicken-Bones** [comment]
Players will change computers over time as well, so mac address isn't good.

I hear your request for something less easily spoofable. However, unless a client obtains another client's uuid, they could only reset their privileges. If your concern is moderation of a public server, then a different solution may be needed there

**Comment by Killia0** [comment]
Yeah I just threw the idea of Mac address out there, certainly not perfect at least on its own and making something reliable isn't simple.
And yes making something secure for public servers it's likely outside the scope of a basic implementation
of what this issue proposes. 
Just thought it was worth mentioning especially for the fact a vanilla uuid can reset on accident which would be annoying for any mod developer relying on it for persistent identification (even when you ignore possibility of intentional manipulation)

**Comment by drunderscore** [comment]
@Killia0
> it is something that can really easily be changed

This is true for any identifying token or UUID clientside ever. On GitHub, you're identified by a cookie, and that can be spoofed easily. I would argue (well, not just me, but every OAuth design since ever) that it's your job to identify invalid UUIDs that have been explicitly spoofed. In the case of impersonation, there's not much you can do, that can happen on the internet too. A naive way to battle this would be to require the connecting IP address to match the one that originally sent the UUID.

Besides it being part of the config JSON (which is a flaw, as it wouldn't be uncommon to want to share your config or delete it to reset it), the UUID is satisfactory -- the spoofing "problem" isn't much a problem as something that you just have to deal with.

**Comment by drunderscore** [comment]
If we want to keep the UUID, it should simply just be moved to it's own file. If we want to go a more "secure" route, then why not do what Mumble does and [use certificates](https://wiki.mumble.info/wiki/Mumble_Certificates)?

This requires no passwords, and can be self-generated when starting the game for the first time, or self supplied by the user to globally identify themselves.

Whatever ends up happening, we should not be reinventing the wheel of user authentication.

**Comment by Killia0** [comment]
> This is true for any identifying token or UUID clientside ever.
Well absolutely there is not such thing as an unspoofable uuid, it's just more difficult to identify when an uuid was spoofed when there is little information to go off on and when it can be changed even on accident.
The format of it being a common GUID is as such not a big deal, just need to consider the config file thing. 
And possibly I guess make it individual per server to not allow impersonating by getting it from someone that joins your server, though that does go again more into the realm of something only really importan for big multiplayer, not just private playthroughs.

**Comment by drunderscore** [comment]
> And possibly I guess make it individual per server to not allow impersonating by getting it from someone that joins your server

I really like this idea! You are right that certificates would be a great choice for this implementation.

**Comment by hamstar0** [comment]
> > The Main.clientUUIID string seems completely satisfactory. Just need to store it somewhere after MessageBuffer receives the packet that contains a player's UUID. It is currently ignored by servers:
> > ![image](https://user-images.githubusercontent.com/6795251/118410749-be117a00-b699-11eb-8434-5a05c7baa668.png)
> 
> Hello, I don't really deal with tmodloader besides recently looking at the github progress to see if I want to delve into it (probably once stable 1.4 is out). But I do deal with servers a lot (On a big vanilla TShock server) so I felt like giving my opinion on this.
> I wouldn't say vanilla `Main.clientUUID` is satisfactory as a reliable way to identify a player; it is something that can really easily be changed (it's just a field on the .json config, with a simple GUID format), which is an issue when either a player wants to be malicious on your server, or just if you rely on this ID for any persistent mapping of values and the player accidentally deleted and reset their config. Ideally you'd have something that absolutely tracks a unique player/game installation that can't be easily spoofed (or well for basic purposes something that wouldn't change on accident) but I'm not sure how you'd want to handle this with steamID not being something every installation has, maybe hashed mac address as an alternative?.

There's nothing that "absolutely tracks a unique player/game installation that can't be easily spoofed" unless you want to get an authentication server (3rd party) involved.

**Comment by Steviegt6** [comment]
If you would allow me to chime in: Steam IDs aren't private information, and are easily obtainable _if you have the profile on hand_. They're only "private" if you don't have any information on the account. The most malicious thing you could do with an ID is change some (mainly) visual stuff regarding an account. Steam IDs still aren't viable because of GOG and manual-installers, though.

**Comment by zzp198-old** [comment]
Let me mention that at present, the SSC mod is used to distinguish users by steamid. I have to give up GoG and normal installation because only steamid is trusted and easy to carry and transfer.
Sad but helpless.

**Comment by zzp198** [comment]
My other alternative method is to add an additional field before connecting to the server to allow the user to input the user name and password by himself, and encrypt the username with the password salt to obtain sha256 as the identity.
If can't get steamid, make one by yourself.
